### ISSUE TITLE
About Sparse-conv module

### ISSUE THREAD

**Issue by Overtown** [Qualcomm-AI-research/Skip-Conv#1]
hi, thanks for your project, I would like to know how to construct sparse-conv that mentioned in your paper?

**Comment by IwakuraRein** [comment]
Seem like this PyTorch project only simulates the skip-conv with the nn.Conv2d. 🤔 

Does this mean I won't have speed improvement from deploying these codes in my project as sparse-conv isn't truly implemented?

**Comment by dabeschte** [comment]
Looks like it. This implementation uses a dense convolution + extra work, so it is even slower than the dense version. This is only a proof of concept

**Comment by IwakuraRein** [comment]
> Looks like it. This implementation uses a dense convolution + extra work, so it is even slower than the dense version. This is only a proof of concept

[This](https://github.com/uber-research/sbnet) seems the implementation of sparse-conv ([paper](https://arxiv.org/abs/1801.02108)). I guess the skip-conv can be achieved  by deploying the NormGate/GumbelGate into it.

**Comment by IwakuraRein** [comment]
I tried [SBNet](https://github.com/uber-research/sbnet)'s lib but weirdly seeing my network is heavily slower. I started an [issue](https://github.com/uber-research/sbnet/issues/28) but since this repo is 2018's, I'm unlikely to get responses there. 😥

Their paper said they implemented skip-conv with img2col based approach from [_More is less: A more complicated network with less inference complexity_](https://scholar.google.com/scholar?q=more+is+less+a+more+complicated+network+with+less+inference+complexity&hl=zh-CN&as_sdt=0&as_vis=1&oi=scholart). I wonder where these codes are released.

**Comment by jiaerwang0328** [comment]
＠IwakuraRein　I've been working on this problem recently, have you found a solution?

**Comment by dabeschte** [comment]
[CBInfer](https://github.com/lukasc-ch/CBinfer) is the only working sparse video framework (that I was able to find) that can actually improve frame rates of some CNNs. However, it does not support batched inference - you will probably only see speedups over cuDNN on low end GPUs because of that.

**Comment by jiaerwang0328** [comment]
@dabeschte 
I followed the semantic segmentation example provided by cbinfer and  found that the time cost is similar:

time_cb: 0.13297295570373535
time_ref: 0.13713407516479492
accuracy (reference): 94.11%
accuracy (CBinfer): 93.88%

Am I doing something wrong? GPU: 2060s, cuda10.0, pytorch: 1.4.0

**Comment by dabeschte** [comment]
yes, the problem here is that CBInfer does not support batched inference out of the box and your GPU is not fully utilized with a batch size of 1 - utilizing less of it does not make inference faster, because you always have to wait on the one core that is actually processing a pixel. 
It can make a difference when you use huge feature maps (I guess mainly with large resolutions).

**Comment by jiaerwang0328** [comment]
@dabeschte I get it, thank you

**Comment by dabeschte** [comment]
@wuranran @IwakuraRein @Overtown 
I developed my own sparse CNN implementation that provides CUDA kernels and a PyTorch integration. The repository is still work in progress (will do more refactoring and add more examples), but I wanted to give you early access since you seem to be looking for a sparse convolution for a while now.
Check out my repository [DeltaCNN](https://github.com/dabeschte/DeltaCNN)
